### ISSUE TITLE
[type hierarchy] Extract some Seq/IndexedSeq methods into a IndexedTraversable interface?

### ISSUE THREAD

**Issue by chronodm** [vavr-io/vavr#2037]
I'm looking at implementing an `ArraySet`, an order-preserving collection with O(1) indexed access, backed by an `Array` and a `HashSet`. Most `IndexedSeq` methods still make reasonable sense for an `IndexedSeq` that allows only unique elements, but some are more problematic if we want them to return the same type:

- `crossProduct()` must either return a different, non-unique type, or drop some combinations
- `intersperse(T)` must either return a different, non-unique type, or be interpreted as `insert(1, element)`
- `padTo()` and `leftPadTo()` have similar problems, and cannot be guaranteed to return a (unique) sequence of the correct size

(Also, the behavior of `insert(int, T)` is somewhat undefined (if the element already exists, should we return `this`, or create a copy that moves the element to the new index?). Likewise `insertAll(int, Iterable<? extends T>)`.)

It would be nice if the methods that don't assume duplicate values are allowed were on a superinterface that could be implemented cleanly.

**Comment by nfekete** [comment]
Just a thought. Maybe it's worth considering splitting the base interfaces of `IndexedSeq` and `Seq` and others to a read-only view interface that doesn't allow any "change" (change in case of immutable collections meaning, of course, returning new instances of the underlying collection type), and another interface for the "change" methods. With that, types like `ArraySet` could be viewed as an indexed sequence, but not manipulated as one, as that doesn't really make sense in the context of `ArraySet`.

**Comment by dmolesUC3** [comment]
Hmm. You could say you can view `ArraySet` as a `Seq` but only manipulate it as a `Set`. But I do think that `append()`, `prepend()` etc. still make sense most of the time, as well as indexed methods like `insert()` and `removeAt()`.

**Comment by nfekete** [comment]
I don't know. What would be the the defined behavior if you would prepend an element into an `ArraySet`, that already contains that element. Same for append or insert? Maybe `removeAt` could make sense, but with all the other `IndexedSeq` methods not being self-evident, I tend to think that it's better not to have them in the first place. `ArraySeq` would be a special collection type, with well-defined behavior for some operations, and not so well-defined for others. Maybe it would be simpler to have just those operations that are self-evident. Having more fine-grained separation of interfaces might help achieving that goal.

**Comment by chb0github** [comment]
A `HashSet` is near O(1) access - the default java implementation is actual O(load factor) which defaults to .75n. `EnumSet` in java definitely is O(1) because it uses the enum ordinal to index an array. The only way you can EVER achieve true O(1) access on a set is if the data is in a fixed domain bound (like a collection of enums). A general purpose hash can't do that but can get close enough

As for your _order preserving_ requirement, I assume you mean _iteration order_ - that's trivially implemented now with Java `LinkedHashSet` backed by a `LinkedList`

So: A vavr implementation would be a `io.vavr.List` + `io.vavr.immutable.HashSet`. The only addition would be to add the "before and after" indices for each entry so you can remove an element and connect the 2 elements in the LinkedList in O(1).

Basically what I am saying is: you're over thinking it. Implement a vavr version of a java `LinkedHashSet` and you're off to the races

**Comment by dmolesUC3** [comment]
@danieldietrich that sounds like a very good solution.

**Comment by dmolesUC3** [comment]
@chb0github I'm not clear on what you're proposing. Access to element _m_ of an _n_-element `HashSet` (or `LinkedHashSet`) isn't O(1), it's O(n).

**Comment by chb0github** [comment]
@dmolesUC3 Unless I am reading something different the [run time for a HashMap](https://docs.oracle.com/javase/8/docs/api/) in plain java is _f * O(n)_ where f is the load factor. The default is .75. This is a [classic implementation](https://en.wikipedia.org/wiki/Hash_table) and is pretty easy to visualize: in an empty hash table any particular entry has a 1/t probability of colliding with a previous entry and thus requiring a resolution phase like mod-prob or linked list. Stands to reason that the more elements you put into a fixed size `HashMap` or set the more likely there is a collision such that it approaches O(N). The entries from wikipedia make it pretty clear that for a uniform, randomly distributed hash, this is amortized O(1)

The `Linked` part of `LinkedHashSet` is for iteration order . if you use the hash table to lookup the nodes to manipulate in the linked list that backs it, you can insert, delete and update in _O(1)_ too. 

Debate this point all you want - refer back to the above links when you do.

All I am saying is that you're overthinking the implementation.

**Comment by chb0github** [comment]
`Set` doesn't not have an index. They are not numbered. So, you don't. Maybe I don't understand what the objective is. Hashing is, by it's nature, not indexable. I mean what would `set.elementAt(1)` retrieve? The element in the table at position 1 assumes

1. The underlying table event _has_ something at position 1
2. Is based on an indexable structure like an array

Are you trying to have a `Set` with a predictable iteration order? That's perfectly reasonable and it's because of this you could _fake_ it. So, if you would expect that `set.elementAt(1) would retrieve the 1st element in an iteration order than that's fine and useful.

**Comment by nfekete** [comment]
@chb0github that's what ArraySet would provide. HashSet combined with Array, to provide (read-only) indexed access of insertion order in O(1).

**Comment by chb0github** [comment]
Ok, now I get what you're after. Maybe just having an interface of `Indexed` and then `IndexedSet` and `List` and `Vector` already implement `Indexed` where as `Set` does not but `IndexedSet` does

**Comment by danieldietrich** [comment]
Seq/Set are distinct Traversable types. We are able to add required methods to Set (that may also be present in Seq) without having a common super-type (like the imaginary IndexedTraversable). We will not introduce additional general collection interfaces (even if non-public).

**Comment by chronodm** [comment]
It will be annoying to have to duplicate code that makes use of the similar methods, but I suppose that would still handle many use cases. Maybe some time around Java SE 15 we'll get proper structural typing.

**Comment by danieldietrich** [comment]
When having proper Java 9 modules it will be a different story. Then (I think, without having verified it) it is possible to hide technical types from reflection. We will see...
